File: android/app/src/main/java/ae/findmybay/feature/bookings/ShowQrViewModel.kt
package ae.findmybay.feature.bookings

import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import ae.findmybay.data.repo.BookingRepository
import ae.findmybay.domain.model.BookingQr
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import javax.inject.Inject

data class ShowQrUiState(
    val loading: Boolean = true,
    val qr: BookingQr? = null,
    val error: String? = null,
)

@HiltViewModel
class ShowQrViewModel @Inject constructor(
    savedStateHandle: SavedStateHandle,
    private val bookings: BookingRepository,
) : ViewModel() {

    val bookingId: String = checkNotNull(savedStateHandle["bookingId"])

    private val _state = MutableStateFlow(ShowQrUiState())
    val state: StateFlow<ShowQrUiState> = _state.asStateFlow()

    init {
        load()
        // Poll every 30s to detect when the vendor scans (status flips to in_progress).
        viewModelScope.launch {
            while (true) {
                delay(30_000)
                if (_state.value.qr?.status != "in_progress") load(silent = true)
            }
        }
    }

    fun load(silent: Boolean = false) {
        if (!silent) _state.update { it.copy(loading = true, error = null) }
        viewModelScope.launch {
            runCatching { bookings.qr(bookingId) }
                .onSuccess { q -> _state.update { it.copy(loading = false, qr = q) } }
                .onFailure { e ->
                    if (!silent) _state.update { it.copy(loading = false, error = e.message ?: "Couldn't load QR") }
                }
        }
    }
}

File: ios/FindMyBay/Features/Bookings/ShowQRViewModel.swift
// ShowQRViewModel.swift
//
// Loads `/bookings/{id}/qr` and polls every 30s so when the vendor scans
// the code (status flips to `in_progress`) we surface that without a
// pull-to-refresh. Mirrors `feature/bookings/ShowQrViewModel.kt`.

import Foundation
import Observation

@Observable @MainActor
final class ShowQRViewModel {
    var loading: Bool = true
    var qr: BookingQr? = nil
    var error: String? = nil

    let bookingId: String
    private let repo: BookingRepository
    private var pollTask: Task<Void, Never>? = nil

    init(bookingId: String, repo: BookingRepository) {
        self.bookingId = bookingId
        self.repo = repo
    }

    func start() {
        load()
        pollTask?.cancel()
        pollTask = Task { [weak self] in
            while let self, !Task.isCancelled {
                try? await Task.sleep(nanoseconds: 30_000_000_000)
                if Task.isCancelled { break }
                if self.qr?.status != "in_progress" {
                    self.load(silent: true)
                }
            }
        }
    }

    func stop() {
        pollTask?.cancel()
        pollTask = nil
    }

    func load(silent: Bool = false) {
        if !silent { loading = true; error = nil }
        Task {
            do {
                qr = try await repo.qr(bookingId: bookingId)
                loading = false
            } catch {
                if !silent {
                    loading = false
                    self.error = (error as? LocalizedError)?.errorDescription
                        ?? "Couldn't load QR"
                }
            }
        }
    }
}
